File: zoo/src/main/scala/com/intel/analytics/zoo/models/seq2seq/Seq2seq.scala
/*
 * Copyright 2018 Analytics Zoo Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.zoo.models.seq2seq

import com.intel.analytics.bigdl._
import com.intel.analytics.bigdl.dataset.{PaddingParam, Sample}
import com.intel.analytics.bigdl.nn.{BatchNormParams, Cell}
import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity}
import com.intel.analytics.bigdl.nn.keras.{KerasLayer, KerasLayerSerializable}
import com.intel.analytics.bigdl.optim.{OptimMethod, ValidationMethod}
import com.intel.analytics.bigdl.serialization.Bigdl.{AttrValue, BigDLModule}
import com.intel.analytics.bigdl.tensor.Tensor
import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
import com.intel.analytics.bigdl.utils._
import com.intel.analytics.bigdl.utils.serializer.converters.DataConverter
import com.intel.analytics.bigdl.utils.serializer.{ContainerSerializable, DeserializeContext, ModuleSerializer, SerializeContext}
import com.intel.analytics.zoo.models.common.ZooModel
import com.intel.analytics.zoo.pipeline.api.keras.models.{KerasNet, Model, Sequential}
import com.intel.analytics.zoo.pipeline.api.keras.layers._
import com.intel.analytics.zoo.pipeline.api.keras.layers.utils.KerasUtils
import org.apache.spark.rdd.RDD

import scala.collection.mutable.ArrayBuffer
import scala.reflect.ClassTag
import scala.reflect.runtime._

/**
 * [[Seq2seq]] A trainable interface for a simple, generic encoder + decoder model
 * @param encoder an encoder object
 * @param decoder a decoder object
 * @param inputShape shape of encoder input, for variable length, please input -1
 * @param outputShape shape of decoder input, for variable length, please input -1
 * @param bridge connect encoder and decoder
 * @param generator Feeding decoder output to generator to generate final result
 */
class Seq2seq[T: ClassTag] (
  val encoder: Encoder[T],
  val decoder: Decoder[T],
  val inputShape: Shape,
  val outputShape: Shape,
  val bridge: KerasLayer[Activity, Activity, T],
  val generator: KerasLayer[Activity, Activity, T])
  (implicit ev: TensorNumeric[T]) extends ZooModel[Table, Tensor[T], T] {

  override def buildModel(): AbstractModule[Table, Tensor[T], T] = {
    val encoderInput = Input(inputShape)
    val decoderInput = Input(outputShape)

    val encoderOutput = encoder.inputs(encoderInput)

    // select table is 0 based
    val encoderFinalStates = SelectTable(1).inputs(encoderOutput)
    val decoderInitStates = if (bridge != null) {
      bridge.inputs(encoderFinalStates)
    }
    else encoderFinalStates

    val decoderOutput = decoder.inputs(Array(decoderInput, decoderInitStates))

    val output = if (generator != null) {
      generator.inputs(decoderOutput)
    }
    else decoderOutput

    Model(Array(encoderInput, decoderInput), output)
      .asInstanceOf[AbstractModule[Table, Tensor[T], T]]
  }

  def compile(
    optimizer: OptimMethod[T],
    loss: Criterion[T],
    metrics: List[ValidationMethod[T]] = null)(implicit ev: TensorNumeric[T]): Unit = {
    model.asInstanceOf[KerasNet[T]].compile(optimizer, loss, metrics)
  }

  def fit(
    x: RDD[Sample[T]],
    batchSize: Int = 32,
    nbEpoch: Int = 10,
    validationData: RDD[Sample[T]] = null,
    featurePaddingParam: PaddingParam[T] = null,
    labelPaddingParam: PaddingParam[T] = null)(implicit ev: TensorNumeric[T]): Unit = {
    model.asInstanceOf[KerasNet[T]].fit(x, batchSize, nbEpoch, validationData,
      featurePaddingParam, labelPaddingParam)
  }

  def setCheckpoint(path: String, overWrite: Boolean = true): Unit = {
    model.asInstanceOf[KerasNet[T]].setCheckpoint(path, overWrite)
  }

  /**
   * Infer output with given input
   * @param input a sequence of data feed into encoder, eg: batch x seqLen x featureSize
   * @param startSign a tensor which represents start and is fed into decoder
   * @param maxSeqLen max sequence length for final output
   * @param stopSign a tensor that indicates model should stop infer further if current
   *                 output is the same with stopSign
   * @param buildOutput Feeding model output to buildOutput to generate final result
   */
  def infer(input: Tensor[T], startSign: Tensor[T], maxSeqLen: Int = 30,
            stopSign: Tensor[T] = null,
            buildOutput: KerasLayer[Tensor[T], Tensor[T], T] = null): Tensor[T] = {
    val sent1 = input
    val sent2 = Tensor[T](startSign.size())
    sent2.copy(startSign)
    sent2.resize(Array(1) ++ startSign.size())

    var curInput = sent2
    val sizes = curInput.size()
    val concat = Tensor[T](Array(sizes(0), maxSeqLen + 1) ++ sizes.drop(2))
    concat.narrow(Seq2seq.timeDim, 1, 1).copy(sent2)
    var break = false

    if (buildOutput != null && !buildOutput.isBuilt()) {
      if (generator != null) {
        buildOutput.build(generator.getOutputShape())
      } else {
        buildOutput.build(decoder.getOutputShape())
      }
    }
    var j = 1
    // Iteratively output predicted words
    while (j <= maxSeqLen && !break) {
      val modelOutput = updateOutput(T(sent1, curInput)).toTensor[T]
      val generateOutput = if (buildOutput != null) buildOutput.forward(modelOutput)
      else modelOutput
      val predict = generateOutput.select(2, generateOutput.size(2))

      if (stopSign != null && predict.almostEqual(stopSign, 1e-8)) break = true
      j += 1
      concat.narrow(Seq2seq.timeDim, j, 1).copy(predict)
      curInput = concat.narrow(Seq2seq.timeDim, 1, j)
    }
    curInput
  }
}

object Seq2seq extends ContainerSerializable {
  ModuleSerializer.registerModule(
    "com.intel.analytics.zoo.models.seq2seq.Seq2seq",
    Seq2seq)

  val timeDim = 2
  /**
   * [[Seq2seq]] A trainable interface for a simple, generic encoder + decoder model
   * @param encoder a rnn encoder object
   * @param decoder a rnn decoder object
   * @param inputShape shape of encoder input, for variable length, please input -1
   * @param outputShape shape of decoder input, for variable length, please input -1
   * @param bridge connect encoder and decoder
   * @param generator Feeding decoder output to generator to generate final result
   */
  def apply[@specialized(Float, Double) T: ClassTag](
    encoder: RNNEncoder[T],
    decoder: RNNDecoder[T],
    inputShape: Shape,
    outputShape: Shape,
    bridge: KerasLayer[Activity, Activity, T] = null,
    generator: KerasLayer[Activity, Activity, T] = null
  )(implicit ev: TensorNumeric[T]): Seq2seq[T] = {
    require(encoder.rnns.length == decoder.rnns.length, "rnn encoder and decoder should has" +
      " the same number of layers!")
    new Seq2seq[T](encoder, decoder, inputShape, outputShape,
      bridge, generator).build()
  }

  /**
   * This factory method is mainly for Python use.
   * Pass in a model to build the Seq2seq.
   * Note that if you use this factory method, arguments such as encoder, decoder, etc
   * should match the model definition to eliminate ambiguity.
   */
  private[zoo] def apply[@specialized(Float, Double) T: ClassTag](
     encoder: RNNEncoder[T],
     decoder: RNNDecoder[T],
     inputShape: Shape,
     outputShape: Shape,
     bridge: KerasLayer[Activity, Activity, T],
     generator: KerasLayer[Activity, Activity, T],
     model: AbstractModule[Table, Tensor[T], T])
   (implicit ev: TensorNumeric[T]): Seq2seq[T] = {
    new Seq2seq[T](encoder, decoder, inputShape, outputShape, bridge, generator)
      .addModel(model)
  }

  /**
   * Load an existing seq2seq model (with weights).
   *
   * @param path The path for the pre-defined model.
   *             Local file system, HDFS and Amazon S3 are supported.
   *             HDFS path should be like "hdfs://[host]:[port]/xxx".
   *             Amazon S3 path should be like "s3a://bucket/xxx".
   * @param weightPath The path for pre-trained weights if any. Default is null.
   * @tparam T Numeric type of parameter(e.g. weight, bias). Only support float/double now.
   */
  def loadModel[T: ClassTag](
    path: String,
    weightPath: String = null)(implicit ev: TensorNumeric[T]): Seq2seq[T] = {
    ZooModel.loadModel(path, weightPath).asInstanceOf[Seq2seq[T]]
  }

  override def doLoadModule[T: ClassTag](context : DeserializeContext)
    (implicit ev: TensorNumeric[T]) : AbstractModule[Activity, Activity, T] = {

    val attrMap = context.bigdlModule.getAttrMap

    val encoderAttr = attrMap.get("encoder")
    val encoder = DataConverter.getAttributeValue(context, encoderAttr).
      asInstanceOf[RNNEncoder[T]]

    val decoderAttr = attrMap.get("decoder")
    val decoder = DataConverter.getAttributeValue(context, decoderAttr).
      asInstanceOf[RNNDecoder[T]]

    val bridgeAttr = attrMap.get("bridge")
    val bridge = DataConverter.getAttributeValue(context, bridgeAttr).
      asInstanceOf[KerasLayer[Activity, Activity, T]]

    val generatorAttr = attrMap.get("generator")
    val generator = DataConverter.getAttributeValue(context, generatorAttr).
      asInstanceOf[KerasLayer[Activity, Activity, T]]

    val inputShapeAttr = attrMap.get("inputShape")
    val inputShape = DataConverter.getAttributeValue(context, inputShapeAttr).asInstanceOf[Shape]

    val outputShapeAttr = attrMap.get("outputShape")
    val outputShape =
      DataConverter.getAttributeValue(context, outputShapeAttr).asInstanceOf[Shape]

    val seq2seq = new Seq2seq(encoder, decoder, inputShape, outputShape, bridge, generator)

    val modelAttr = attrMap.get("model")
    val model = DataConverter.getAttributeValue(context, modelAttr).
      asInstanceOf[AbstractModule[Table, Tensor[T], T]]
    seq2seq.addModel(model)
    seq2seq
  }

  override def doSerializeModule[T: ClassTag](context: SerializeContext[T],
    seq2seqBuilder : BigDLModule.Builder)
    (implicit ev: TensorNumeric[T]) : Unit = {

    val seq2seq = context.moduleData.module.asInstanceOf[Seq2seq[T]]

    seq2seq.encoder.asInstanceOf[RNNEncoder[T]].inputShape = seq2seq.inputShape
    val encoderBuilder = AttrValue.newBuilder
    DataConverter.setAttributeValue(context, encoderBuilder,
      seq2seq.encoder, ModuleSerializer.abstractModuleType)
    seq2seqBuilder.putAttr("encoder", encoderBuilder.build)

    val statesShape = if (seq2seq.bridge != null) {
      KerasUtils.removeBatch(seq2seq.bridge.getOutputShape())
    } else KerasUtils.removeBatch(Shape(seq2seq.encoder.getOutputShape().toMulti().drop(1)))
    val decoderShape =
      MultiShape(List(KerasUtils.removeBatch(seq2seq.encoder.getOutputShape().toMulti().head),
        statesShape))
    seq2seq.decoder.asInstanceOf[RNNDecoder[T]].inputShape = decoderShape
    val decoderBuilder = AttrValue.newBuilder
    DataConverter.setAttributeValue(context, decoderBuilder,
      seq2seq.decoder, ModuleSerializer.abstractModuleType)
    seq2seqBuilder.putAttr("decoder", decoderBuilder.build)

    val bridgeBuilder = AttrValue.newBuilder
    DataConverter.setAttributeValue(context, bridgeBuilder,
      seq2seq.bridge, ModuleSerializer.abstractModuleType)
    seq2seqBuilder.putAttr("bridge", bridgeBuilder.build)

    val generatorBuilder = AttrValue.newBuilder
    DataConverter.setAttributeValue(context, generatorBuilder,
      seq2seq.generator, ModuleSerializer.abstractModuleType)
    seq2seqBuilder.putAttr("generator", generatorBuilder.build)

    val shapeBuilder = AttrValue.newBuilder
    DataConverter.setAttributeValue(context, shapeBuilder,
      seq2seq.inputShape, universe.typeOf[Shape])
    seq2seqBuilder.putAttr("inputShape", shapeBuilder.build)

    val outputShapeBuilder = AttrValue.newBuilder
    DataConverter.setAttributeValue(context, outputShapeBuilder,
      seq2seq.outputShape, universe.typeOf[Shape])
    seq2seqBuilder.putAttr("outputShape", outputShapeBuilder.build)

    val laborBuilder = AttrValue.newBuilder
    DataConverter.setAttributeValue(context, laborBuilder,
      seq2seq.model, ModuleSerializer.abstractModuleType)
    seq2seqBuilder.putAttr("model", laborBuilder.build)
  }
}

File: pyzoo/zoo/models/seq2seq/seq2seq.py
#
# Copyright 2018 Analytics Zoo Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from bigdl.nn.layer import Layer
from zoo.pipeline.api.keras.layers import *
from zoo.models.common import ZooModel

from zoo.pipeline.api.keras.engine import ZooKerasLayer
from zoo.pipeline.api.keras.models import Model


if sys.version >= '3':
    long = int
    unicode = str


def createRNN(rnn_type, nlayers, hidden_size):
    _rnn_type = rnn_type.lower()
    if (_rnn_type == "lstm"):
        return [LSTM(hidden_size, return_sequences=True) for layer in range(nlayers)]
    elif (_rnn_type == "gru"):
        return [GRU(hidden_size, return_sequences=True) for layer in range(nlayers)]
    elif (_rnn_type == "simplernn"):
        return [SimpleRNN(hidden_size, return_sequences=True) for layer in range(nlayers)]
    else:
        raise Exception('Only support lstm|gru|simplernn')


class RNNEncoder(ZooKerasLayer):
    """
    A generic recurrent neural network encoder

    # Arguments
    rnns: rnn layers used for encoder, support stacked rnn layers
    embedding: embedding layer in encoder
    input_shape: shape of input, not including batch

    >>> encoder = RNNEncoder.initialize("lstm", 2, 3)
    creating: createZooKerasLSTM
    creating: createZooKerasLSTM
    creating: createZooKerasRNNEncoder

    >>> lstm = LSTM(3)
    creating: createZooKerasLSTM
    >>> embedding = Embedding(1000, 32, input_length=10, name="embedding1")
    creating: createZooKerasEmbedding
    >>> encoder = RNNEncoder([lstm], embedding)
    creating: createZooKerasRNNEncoder
    """
    def __init__(self, rnns, embedding=None, input_shape=None):
        super(RNNEncoder, self).__init__(None,
                                         rnns,
                                         embedding,
                                         list(input_shape) if input_shape else None)

    @classmethod
    def initialize(cls, rnn_type, nlayers, hidden_size, embedding=None, input_shape=None):
        """
        rnn_type: currently support "simplernn | lstm | gru"
        nlayers: number of layers used in encoder
        hidden_size: hidden size of encoder
        embedding: embedding layer in encoder, `None` is supported
        """
        rnns = createRNN(rnn_type, nlayers, hidden_size)
        return RNNEncoder(rnns, embedding, input_shape)


class RNNDecoder(ZooKerasLayer):
    """
    A generic recurrent neural network decoder

    # Arguments
    rnns: rnn layers used for decoder, support stacked rnn layers
    embedding: embedding layer in decoder
    input_shape: shape of input, not including batch

    >>> decoder = RNNDecoder.initialize("lstm", 2, 3)
    creating: createZooKerasLSTM
    creating: createZooKerasLSTM
    creating: createZooKerasRNNDecoder

    >>> lstm = LSTM(3)
    creating: createZooKerasLSTM
    >>> embedding = Embedding(1000, 32, input_length=10, name="embedding1")
    creating: createZooKerasEmbedding
    >>> encoder = RNNDecoder([lstm], embedding)
    creating: createZooKerasRNNDecoder
    """
    def __init__(self, rnns, embedding=None, input_shape=None):
        super(RNNDecoder, self).__init__(None,
                                         rnns,
                                         embedding,
                                         list(input_shape) if input_shape else None)

    @classmethod
    def initialize(cls, rnn_type, nlayers, hidden_size, embedding=None, input_shape=None):
        """
        rnn_type: currently support "simplernn | lstm | gru"
        nlayers: number of layers used in decoder
        hidden_size: hidden size of decoder
        embedding: embedding layer in decoder, `None` is supported
        """
        rnns = createRNN(rnn_type, nlayers, hidden_size)
        return RNNDecoder(rnns, embedding, input_shape)


class Bridge(ZooKerasLayer):
    """
    defines how to transform encoder to decoder

    # Arguments
    bridge_type: currently only support "dense | densenonlinear"
    decoder_hiddenSize: hidden size of decoder
    bridge: keras layers used to do the transformation

    >>> bridge = Bridge.initialize("dense", 2)
    creating: createZooKerasBridge
    >>> dense = Dense(3)
    creating: createZooKerasDense
    >>> bridge = Bridge.initialize_from_keras_layer(dense)
    creating: createZooKerasBridge
    """
    def __init__(self, bridge_type, decoder_hidden_size, bridge):
        super(Bridge, self).__init__(None, bridge_type, decoder_hidden_size, bridge)

    @classmethod
    def initialize(cls, bridge_type, decoder_hidden_size):
        """
        bridge_type: currently only support "dense | densenonlinear"
        decoder_hiddenSize: hidden size of decoder
        """
        return Bridge(bridge_type, decoder_hidden_size, None)

    @classmethod
    def initialize_from_keras_layer(cls, bridge):
        """
        bridge: keras layers used to do the transformation
        """
        return Bridge("customized", 0, bridge)


class Seq2seq(ZooModel):
    """
    A trainable interface for a simple, generic encoder + decoder model

    # Arguments
    encoder: an encoder object
    decoder: a decoder object
    input_shape: shape of encoder input, for variable length, please use -1 as seq len
    output_shape: shape of decoder input, for variable length, please use -1 as seq len
    bridge: connect encoder and decoder
    generator: Feeding decoder output to generator to generate final result, `None` is supported

    >>> encoder = RNNEncoder.initialize("LSTM", 1, 4)
    creating: createZooKerasLSTM
    creating: createZooKerasRNNEncoder
    >>> decoder = RNNDecoder.initialize("LSTM", 1, 4)
    creating: createZooKerasLSTM
    creating: createZooKerasRNNDecoder
    >>> bridge = Bridge.initialize("dense", 4)
    creating: createZooKerasBridge
    >>> seq2seq = Seq2seq(encoder, decoder, [2, 4], [2, 4], bridge)
    creating: createZooKerasInput
    creating: createZooKerasInput
    creating: createZooKerasSelectTable
    creating: createZooKerasModel
    creating: createZooSeq2seq
    """

    def __init__(self, encoder, decoder, input_shape, output_shape, bridge=None,
                 generator=None, bigdl_type="float"):
        if (input_shape is None) or (output_shape is None):
            raise TypeError('input_shape and output_shape cannot be None')
        self.encoder = encoder
        self.decoder = decoder
        self.input_shape = list(input_shape)
        self.output_shape = list(output_shape)
        self.bridge = bridge
        self.generator = generator
        self.bigdl_type = bigdl_type
        self.model = self.build_model()
        super(Seq2seq, self).__init__(None, self.bigdl_type,
                                      self.encoder,
                                      self.decoder,
                                      self.input_shape,
                                      self.output_shape,
                                      self.bridge,
                                      self.generator,
                                      self.model)

    def build_model(self):
        encoder_input = Input(name="encoder_input", shape=self.input_shape)
        decoder_input = Input(name="decoder_input", shape=self.output_shape)
        encoder_output = self.encoder(encoder_input)

        encoder_final_states = SelectTable(1)(encoder_output)
        decoder_init_states =\
            self.bridge(encoder_final_states) if self.bridge else encoder_final_states

        decoder_output = self.decoder([decoder_input, decoder_init_states])

        output = self.generator(decoder_output) if self.generator else decoder_output

        return Model([encoder_input, decoder_input], output)

    def set_checkpoint(self, path, over_write=True):
        callBigDlFunc(self.bigdl_type, "seq2seqSetCheckpoint",
                      self.value,
                      path,
                      over_write)

    @staticmethod
    def load_model(path, weight_path=None, bigdl_type="float"):
        """
        Load an existing Seq2seq model (with weights).

        # Arguments
        path: The path for the pre-defined model.
              Local file system, HDFS and Amazon S3 are supported.
              HDFS path should be like 'hdfs://[host]:[port]/xxx'.
              Amazon S3 path should be like 's3a://bucket/xxx'.
        weight_path: The path for pre-trained weights if any. Default is None.
        """
        jmodel = callBigDlFunc(bigdl_type, "loadSeq2seq", path, weight_path)
        model = ZooModel._do_load(jmodel, bigdl_type)
        model.__class__ = Seq2seq
        return model

    # For the following methods, please refer to KerasNet for documentation.
    def compile(self, optimizer, loss, metrics=None):
        if isinstance(optimizer, six.string_types):
            optimizer = to_bigdl_optim_method(optimizer)
        if isinstance(loss, six.string_types):
            loss = to_bigdl_criterion(loss)
        if metrics and all(isinstance(metric, six.string_types) for metric in metrics):
            metrics = to_bigdl_metrics(metrics, loss)
        callBigDlFunc(self.bigdl_type, "seq2seqCompile",
                      self.value,
                      optimizer,
                      loss,
                      metrics)

    def fit(self, x, batch_size=32, nb_epoch=10, validation_data=None):
        callBigDlFunc(self.bigdl_type, "seq2seqFit",
                      self.value,
                      x,
                      batch_size,
                      nb_epoch,
                      validation_data)

    def infer(self, input, start_sign, max_seq_len=30, stop_sign=None, build_output=None):
        """
        Inference API for given input

        # Arguments
        input: a sequence of data feed into encoder, eg: batch x seqLen x featureSize
        start_sign: a ndarray which represents start and is fed into decoder
        max_seq_len: max sequence length for final output
        stop_sign: a ndarray that indicates model should stop infer further if current output
        is the same with stopSign
        build_output: Feeding model output to buildOutput to generate final result
        """
        jinput, input_is_table = Layer.check_input(input)
        assert not input_is_table
        jstart_sign, start_sign_is_table = Layer.check_input(start_sign)
        assert not start_sign_is_table
        if stop_sign:
            jstop_sign, stop_sign_is_table = Layer.check_input(stop_sign)
            assert not start_sign_is_table
        else:
            jstop_sign = None
        results = callBigDlFunc(self.bigdl_type, "seq2seqInfer",
                                self.value,
                                jinput[0],
                                jstart_sign[0],
                                max_seq_len,
                                jstop_sign[0] if jstop_sign else None,
                                build_output)
        return results
